src/main/java/com/parasoft/em/client/impl/JobsImpl.java

```diff
@@ -124,15 +127,38 @@ public boolean monitorExecution(JSONObject history, EventMonitor monitor) throws
 		} else {
 			result = false;
 		}
+		Map<Long, JSONArray> historyMap = new TreeMap<Long, JSONArray>();
+		JSONArray testHistories = history.optJSONArray("testHistories");
+		if (testHistories != null) {
+			for (int i = 0; i < testHistories.size(); i++) {
+				JSONObject obj = testHistories.getJSONObject(i);
+				Long reportId = obj.getLong("reportId");
+				if (!historyMap.containsKey(reportId)) {
+					historyMap.put(reportId, new JSONArray());
+				}
+				JSONArray tests = historyMap.get(reportId);
```

Comment:
You could optimize this loop which currently does at least two map lookups for each test.
Change map.conatinsKey() to map.get() and then check if the result is null.